plugins/inputs/intel_powerstat/services/msr.go

```diff
@@ -0,0 +1,297 @@
+func (m *MsrServiceImpl) GetTimestampDelta(cpuID string) uint64 {
+	return m.cpuCoresData[cpuID].TscDelta
+}
+
+// GetMperfDelta returns mperf delta for CPU.
+func (m *MsrServiceImpl) GetMperfDelta(cpuID string) uint64 {
+	return m.cpuCoresData[cpuID].MperfDelta
+}
+
+// GetAperfDelta returns aperf delta for CPU.
+func (m *MsrServiceImpl) GetAperfDelta(cpuID string) uint64 {
+	return m.cpuCoresData[cpuID].AperfDelta
+}
+
+// GetC6Delta returns C6 delta for CPU.
+func (m *MsrServiceImpl) GetC6Delta(cpuID string) uint64 {
+	return m.cpuCoresData[cpuID].C6Delta
+}
+
+// GetC3Delta returns C3 delta for CPU.
+func (m *MsrServiceImpl) GetC3Delta(cpuID string) uint64 {
+	return m.cpuCoresData[cpuID].C3Delta
+}
+
+// GetC7Delta returns C7 delta for CPU.
+func (m *MsrServiceImpl) GetC7Delta(cpuID string) uint64 {
+	return m.cpuCoresData[cpuID].C7Delta
+}
+
+// GetReadDate returns last read date.
+func (m *MsrServiceImpl) GetReadDate(cpuID string) int64 {
+	return m.cpuCoresData[cpuID].ReadDate
+}
+
+// SetReadDate sets last read date.
+func (m *MsrServiceImpl) SetReadDate(cpuID string, date int64) {
+	m.cpuCoresData[cpuID].ReadDate = date
+}
+
+// OpenAndReadMsr open and reads data from MSR.
+func (m *MsrServiceImpl) OpenAndReadMsr(core string) error {
+	path := fmt.Sprintf(msrPartialPath, core)
+	msrFile, err := os.Open(path)
+	if err != nil {
+		return fmt.Errorf("error opening MSR file on path %s, err: %v", path, err)
+	}
+	defer msrFile.Close()
+
+	err = m.readDataFromMsr(core, msrFile)
+	if err != nil {
+		return fmt.Errorf("error reading data from MSR for core %s, err: %v", core, err)
+	}
+	return nil
+}
+
+// RetrieveCPUFrequencyForCore retrieves CPU frequency for given core.
+func (m *MsrServiceImpl) RetrieveCPUFrequencyForCore(core string) (float64, int64, error) {
+	cpuFreqPath := fmt.Sprintf(cpuCurrentFreqPartialPath, core)
+	cpuFreqFile, err := os.Open(cpuFreqPath)
+	if err != nil {
+		return 0, 0, fmt.Errorf("error opening scaling_cur_freq file on path %s, err: %v", cpuFreqPath, err)
+	}
+	defer cpuFreqFile.Close()
+	return m.fs.ReadFileToFloat64(cpuFreqFile)
+}
+
+func (m *MsrServiceImpl) readDataFromMsr(core string, reader io.ReaderAt) error {
+	g, ctx := errgroup.WithContext(context.Background())
+
+	// Create and populate a map that contains msr offsets along with their respective channels
+	msrOffsetsWithChannels := make(map[string]chan uint64)
+	for offset := range m.msrOffsets {
+		msrOffsetsWithChannels[offset] = make(chan uint64)
+	}
+
+	// Start a goroutine for each msr offset
+	for offset, channel := range msrOffsetsWithChannels {
+		// Wrap around function to avoid race on loop counter
+		func(off string, ch chan uint64) {
+			g.Go(func() error {
+				defer close(ch)
+
+				err := m.readValueFromFileAtOffset(ctx, ch, reader, m.msrOffsets[off])
+				if err != nil {
+					return fmt.Errorf("error reading MSR file, err: %v", err)
+				}
+
+				return nil
+			})
+		}(offset, channel)
+	}
```

Comment:
What is the rational behind parallelizing this? All data is read from the same file right?

Reply:
Yes, we read 8 bytes at different offsets from the same "character device file" - MSR. And it is faster to do that in parallel.